src/cat-data-service/src/types/proposal.rs

```diff
@@ -0,0 +1,281 @@
+use super::SerdeType;
+use event_db::types::event::proposal::{
+    Proposal, ProposalDetails, ProposalId, ProposalSummary, ProposerDetails,
+};
+use serde::{
+    de::Deserializer,
+    ser::{SerializeStruct, Serializer},
+    Deserialize, Serialize,
+};
+
+impl Serialize for SerdeType<&ProposalId> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        self.0.serialize(serializer)
+    }
+}
+
+impl Serialize for SerdeType<ProposalId> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        SerdeType(&self.0).serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for SerdeType<ProposalId> {
+    fn deserialize<D>(deserializer: D) -> Result<SerdeType<ProposalId>, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Ok(ProposalId(i32::deserialize(deserializer)?).into())
+    }
+}
+
+impl Serialize for SerdeType<&ProposalSummary> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut serializer = serializer.serialize_struct("ProposalSummary", 3)?;
+        serializer.serialize_field("id", &SerdeType(&self.id))?;
+        serializer.serialize_field("title", &self.title)?;
+        serializer.serialize_field("summary", &self.summary)?;
+        serializer.end()
+    }
+}
+
+impl Serialize for SerdeType<ProposalSummary> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        SerdeType(&self.0).serialize(serializer)
+    }
+}
+
+impl Serialize for SerdeType<&ProposerDetails> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut serializer = serializer.serialize_struct("ProposerDetails", 4)?;
```

Comment:
Do we have any options not to use dynamic dispatch? We know the shape of the data in compilation time, right?

Reply:
Not sure that it is possible, because it is the only way how to define serializations for structs.
Also if deeply take a look under the `derive(Serialize)` implementation of serde we will the same approach
https://github.com/serde-rs/serde/blob/992a01bad2c65c2920296cdb0d6708bdd78eae4c/serde_derive/src/ser.rs#L350C1-L351C1

Reply:
Or what do you mean here ?

Reply:
Discussed offline.